### PR TITLE
Port Timer to Goodies

### PR DESCRIPTION
**Pull request by GuiltyDolphin**
Port the Timer Spice to Goodies as there were no API calls involved.
### Test Changes
- Moved the query functions to Perl and added tests for the detection of starting times.

Spice PR: duckduckgo/zeroclickinfo-spice#2491

Porting Issue: duckduckgo/zeroclickinfo-spice#2463

---

IA Page: https://duck.co/ia/view/timer_goodie

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by GuiltyDolphin** [comment]
@moollaza The IA page will need to be updated so the meta-data checks pass.

Apart from that, I'm done porting so we should deploy to beta to make sure there are no regressions.

File: lib/DDG/Goodie/Timer.pm [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                itemType   => 'timer',
+            },
+            data => {
+                time => "$time",
+            },
+            templates => {
+                group       => 'base',
+                detail      => 'DDH.timer.timer_wrapper',
+                wrap_detail => 'base_detail',
+            },
```
Comment (pnpninja) [inline lib/DDG/Goodie/Timer.pm]: That comma could cause trouble

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timer.pm]: Shouldn't, this is Perl - trailing commas is fine (and I like them because they make diffs easier to read!)

**Comment by GuiltyDolphin** [comment]
@moollaza Ping! The IA page needs updating so the tests can pass and we can get this one Beta. Need to check if there's any regressions.

**Comment by moollaza** [comment]
> The IA page needs updating so the tests can pass and we can get this one Beta. Need to check if there's any regressions.

@GuiltyDolphin Let's make a temporary IA Page for this -- I'm not sure if switching a live IA's type will cause any problems.

@jdorweiler @MariagraziaAlastra any ideas? We're porting Spice IAs that are pure JS to Goodies, so to deploy them the IA Page needs to indicate they're Goodies, but the production version of still a Spice...

**Comment by GuiltyDolphin** [comment]
Ping @moollaza @MariagraziaAlastra Tests are passing now - want to try deploying again?

**Comment by GuiltyDolphin** [comment]
Ping @moollaza @zachthompson Situation? Able to deploy?

**Comment by jdorweiler** [comment]
@moollaza sorry just saw this.  A temp IA page should be fine for testing.  I'm not sure about how to actually get it deployed though.  Would there be any problem with merging this and once it's deployed change the repo on the IA page to goodies?

**Comment by GuiltyDolphin** [comment]
@moollaza @jdorweiler I switched it over to https://duck.co/ia/view/timer_goodie a while back - if any case surely we could just update all the info in the new IA page to match the old one? Then remove the other one when this is live?

**Comment by jdorweiler** [comment]
@moollaza @GuiltyDolphin sounds like this is done then?

**Comment by GuiltyDolphin** [comment]
@jdorweiler As far as I know - yes; I haven't noticed any regressions.

@mattr555 you happy with this?

**Comment by mattr555** [comment]
This looks like it's ready to go to me. @moollaza @jdorweiler Please merge and then advise how to continue with the other IAs to port, thanks :)

File: lib/DDG/Goodie/Timer.pm [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+zci answer_type => 'timer';
+zci is_cached   => 1;
+
+my @triggers = qw(timer countdown alarm);
+# Triggers that are valid in start only
+my @startTringgers = qw(start begin set run);
+# Beautifies the trigger can be appended in front/back of trigger
+my @beautifierTringgers = qw(online);
+#Joins the Timer Value
+my @joiners = qw(for on at with);
```
Comment (riqpe) [inline lib/DDG/Goodie/Timer.pm]: @GuiltyDolphin would you consider adding 'in' as another joiner, as in "alarm in 30 minutes"

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timer.pm]: I wouldn't mind, @mattr555 you happy with that?

Comment (mattr555) [inline lib/DDG/Goodie/Timer.pm]: Fine by me 👍

**Comment by tagawa** [comment]
I've just installed the latest version of this on the beta server for final testing: https://beta.duckduckgo.com/?q=timer%205s

**Comment by GuiltyDolphin** [comment]
Thanks @tagawa.

@moollaza Looks like the old one is still on Beta? The new triggers don't seem to be working.

**Comment by mattr555** [comment]
What's the status on this port? Would love to get this out so it doesn't diverge from the one in spice, where there are a few issues that need attention.

**Comment by moollaza** [comment]
@GuiltyDolphin can you resolve the conflicts on your Spice PR again? I'll be able to deploy these to Beta then to confirm it's all good so we can move it along

**Comment by GuiltyDolphin** [comment]
@moollaza Done (and I've updated this branch to include the changes that caused the conflict).

**Comment by moollaza** [comment]
@GuiltyDolphin whoops, I forgot that I need you to use a new ID for this so I can deploy it. I can't change the live metadata without causing problems, so in order to get this on beta it needs to use a new ID. Once we confirm it's working correctly, you'll need to switch the ID back to `timer`, and then I can merge + deprecate the Spice when this code goes live.

Sorry for the extra work!

**Comment by GuiltyDolphin** [comment]
@moollaza No worries!

I've updated the ID to `timer_goodie`, though it doesn't appear to be working correctly locally. Something I missed?

**Comment by moollaza** [comment]
> though it doesn't appear to be working correctly locally (not displaying). Something I missed?

@GuiltyDolphin I think it's related to a bug in the way I had setup the template compiling JS, for Spices. If you change Goodie/Timer.pm to TimerGoodie.pm (to match the ID) it works, however I'm seeing a JS error once the timer executes...

**Comment by moollaza** [comment]
I've update the IA Page accordingly already 👍

**Comment by GuiltyDolphin** [comment]
@moollaza Ah, yep. It's working now, thanks!

File: share/goodie/timer_goodie/timer_goodie.js [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+
+            // create new timer and insert it before the add button
+            addTimer(0);
+        });
+    }
+    return {
+        //wait for the Goodie to load before displaying things
+        //this makes sure the divs display at the right time so the layout doesn't break
+        onShow: onShow,
+        data: {
```
Comment (mattr555) [inline share/goodie/timer_goodie/timer_goodie.js]: I have to comment this out to get it to auto-fill a time from the query

Comment (GuiltyDolphin) [inline share/goodie/timer_goodie/timer_goodie.js]: Alright, I've commented it out.

File: share/goodie/timer_goodie/timer_goodie.js [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+License: CC BY-NC 3.0 http://creativecommons.org/licenses/by-nc/3.0/
+*/
+
+DDH.timer_goodie = DDH.timer_goodie || {};
+
+DDH.timer_goodie.build = function(ops) {
+    'use strict';
+
+    var SOUND_NAME = "alarm-sound",
+        soundUrl = '/share/goodie/timer_goodie/alarm.mp3',
```
Comment (mattr555) [inline share/goodie/timer_goodie/timer_goodie.js]: @moollaza does this need to be versioned (as in #3500)?

Comment (GuiltyDolphin) [inline share/goodie/timer_goodie/timer_goodie.js]: @mattr555 It looks like the changes for that issue aren't merged yet, so we can probably wait and move that PR over to Goodies.

Comment (moollaza) [inline share/goodie/timer_goodie/timer_goodie.js]: > does this need to be versioned (as in #3500)?

yes, any `/share/` assets accessed from the frontend need to use versioned paths. We still need to expose a function for creating those paths for Goodies...

**Comment by moollaza** [comment]
@GuiltyDolphin sorry for the delay, this is deployed and working 👍 

I've updated the Tab name on the IA Page to "Answer" from "Cheat Sheet"

We need to pass thee Goodie Version number from the Perl to the JS to build the asset path and then this should be mergeable.

We do that in a few Goodies, like the solar system: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/SolarSystem.pm#L14

**Comment by moollaza** [comment]
Merging. I will add the audio assets 👍

**Pull request merged**